### PR TITLE
Jkamrx/fix node api remove datafile

### LINKED ISSUE #2700: Metadata/Data File Editing for Existing DataSets
Its a common use case for users to be able to edit errors in their metadata, and upload and associate new data files or replace existing data files after a successful metadata upload.

HSCI Scenarios:

- 1: An analyst uploads an ISA archive and corresponding data files but needs to correct an error in metadata
  - Conditions:
    - No files can be added or removed. Only typographical errors or errors in the content of the metadata are allowed to be edited.
    - Only ISATabs will be supported in this iteration. Tabular metadata upload requires some extra user intervention i.e. selection of `Data File Column`.
    - No `Analyses` or `Visualizations` can be performed on the DataSet
    - Data files uploaded before the edits have been made will be reassociated with the new `DataSet`

- 2: An analyst successfully uploads an ISA archive but needs to add some data files later
  - Functionality will be added to the File Browser to be able to upload and associate files with existing `Nodes` as well as provide a valid url to a data file.
    - This will potentially mean overwriting/purging existing data files so we should be wary of that
  - This functionality will exist for all `DataSets` regardless if they are ISATab or Tabular

### ToDos:
- Frontend
  - [x] Button to `Update DataSet` in the FileBrowser dependent on aforementioned conditions
  - ~~~[ ] Modal to allow selection of new ISATab metadatafile~~~ Redirects to import pages now
  - [x] Extend File Browser code to be able to upload and associate new data files with existing `Nodes`
- Backend
  - [x] Extend API endpoints for `DataSet` update
    - [x] DataSets should be able to report on their "cleanliness"  Ref: #2784
    - ~~~[ ] Check that newly referenced data files are a 1::1 match with what existed prior~~~ We ended up vetoing this and allowing removal and addition of datafiles during a metadata revision
    - [x] Reassociate data files with updated `DataSet` if referenced in new metadata
    - [x] `DataSet` version should be incremented

Add/remove/replace data files for any existing data set that `is_clean`

### PR DESCRIPTION
**Pull request by jkmarx**
Resolves #2700 
During local manual tests:
- Issues with removing & adding files
  Fixed: We should not remove the node's file_uuid. It needs the file_store_item uuid to future uploads. 1) Keep delete datafile 2) Remove import_task_id -> update solr index to 'N/A'

@hackdna @scottx611x Since we're not updating the node object, should we move this to a file_store_item v2 api? I'd remove the node v2 api all together, since we're not using it.
@jkmarx The short answer is probably yes, the long answer is here: https://github.com/refinery-platform/refinery-platform/pull/2854#issuecomment-404239106

- Added datafile index to prevent remove for remote download datafiles
screen shot 2018-07-12 at 10 52 37 pm

@hackdna Should we do anything with the celery task object? 
@jkmarx `FileStoreItem.delete_datafile()` takes care of that

### CODE REVIEW

**Comment by scottx611x** [comment]
I think it makes sense to keep this functionality in a `Node` API. The mentioned `FileStoreItem` v2 API would still have to know about a `Node` to update its corresponding Solr index entry. I think that, given the hierarchy, its more appropriate for `Nodes` to know about and act upon `FileStoreItems` rather than the other way around.

**Comment by hackdna** [comment]
The issue is that the existing data model is not designed to support file removal. Here's an outline for a fix:

* replace `file_uuid` attribute of `Node` with a foreign key to `FileStoreItem`: `Node.data_file = ForeignKey(FileStoreItem, null=True, unique=True, on_delete=SET_NULL)`
* add Solr update logic to `Node.save()`
* move the `source` attribute from `FileStoreItem` to `Node`

This would make file operations simpler and more logical:

Removing data files: `DELETE` request to `FileStoreItem` API. This will automatically stop the file import task (if any) then set `Node.data_file` to null and update Solr when the corresponding `Node` instance is saved after update

Adding data files:` POST` request to `AddFilesToDataSetView`
This will be very similar to the existing implementation: iterate over file `Node` instances in the data set, create corresponding `FileStoreItem` instances and start file import tasks
or
for each new file: `PATCH` request to Node API (to create a new FileStoreItem instance, associate with the `Node` instance and start file import task)

This also makes it easier to move files from one Node to another in case of metadata updates (simply associate the `FileStoreItem` instance with a different `Node`) and  would allow for much simplified `file_store` app operations in general.

**Comment by scottx611x** [comment]
This sounds like a good approach, but I think there are some things to consider.

When manipulating the fields on `FileStoreItem` and `Nodes` there will need to be some thought into the data migrations that need to be introduced to update existing `Nodes` and `FileStoreItems` in our system.

The majority of `Nodes` don't encapsulate files, so it wasn't clear to me why a `source` field would move there.

Overall, this seems like quite a bit of work to tack on to the end of this milestone, but I could be wrong.

File: refinery/core/views.py [hunk 1]
```diff
@@ -754,20 +755,16 @@ def get(self, request, uuid):
...
-                return Response(
-                    serializer.data, status=status.HTTP_202_ACCEPTED
-                )
+            # to remove the data file, we need to delete it and update index,
+            #  the file store item uuid should remain
+            self.update_file_store(node.file_uuid, new_file_uuid)
+            NodeIndex().update_object(node, using="data_set_manager")
             return Response(
-                serializer.errors, status=status.HTTP_400_BAD_REQUEST
+                NodeSerializer(node).data, status=status.HTTP_202_ACCEPTED
```
Comment (hackdna) [inline refinery/core/views.py]: All the operations above are synchronous, so HTTP 200 status is appropriate here.

File: refinery/core/views.py [hunk 2]
```diff
@@ -728,7 +728,8 @@ def update_file_store(self, old_uuid, new_uuid):
             except (FileStoreItem.DoesNotExist,
                     FileStoreItem.MultipleObjectsReturned) as e:
                 logger.error(e)
-            file_store_item.delete_datafile()
+            if file_store_item:
```
Comment (hackdna) [inline refinery/core/views.py]: `else:` is more appropriate here

**Comment by hackdna** [comment]
Data migrations would simply need to have some extra code to copy and update information in the corresponding fields.

The source should be a part of Node to avoid losing that information when a corresponding FileStoreItem instance is deleted. Also, there is already some of file-related functionality in the Node model, so adding a source field is not really an issue. Alternatively, there can be a FileNode model derived from Node that would encapsulate file-related functionality.

This is a bit of work but it is likely we will keep running into this limitation down the road.

File: refinery/core/views.py [hunk 3]
```diff
@@ -754,20 +755,16 @@ def get(self, request, uuid):
...
-                self.update_file_store(old_file_uuid, node.file_uuid)
-                NodeIndex().update_object(node, using="data_set_manager")
-                serializer.save()
-                return Response(
-                    serializer.data, status=status.HTTP_202_ACCEPTED
-                )
+            # to remove the data file, we need to delete it and update index,
+            #  the file store item uuid should remain
+            self.update_file_store(node.file_uuid, new_file_uuid)
+            NodeIndex().update_object(node, using="data_set_manager")
```
Comment (hackdna) [inline refinery/core/views.py]: This statement is now repeated several times throughout the codebase and thus should be moved to `Node.save()` or similar.

File: refinery/core/views.py [hunk 4]
```diff
@@ -754,20 +755,16 @@ def get(self, request, uuid):
...
-            if serializer.is_valid():
-                self.update_file_store(old_file_uuid, node.file_uuid)
-                NodeIndex().update_object(node, using="data_set_manager")
-                serializer.save()
-                return Response(
-                    serializer.data, status=status.HTTP_202_ACCEPTED
-                )
+            # to remove the data file, we need to delete it and update index,
+            #  the file store item uuid should remain
+            self.update_file_store(node.file_uuid, new_file_uuid)
```
Comment (hackdna) [inline refinery/core/views.py]: If this is a method on Node model then it could also handle Solr index update.

File: refinery/core/views.py (hunk 1 above)
Comment (hackdna) [reply refinery/core/views.py]: Also, is `file_uuid` actually updated in the database?

Comment (jkmarx) [reply refinery/core/views.py]: @hackdna No, see pull request comments above.

File: refinery/core/views.py (hunk 3 above)
Comment (jkmarx) [reply refinery/core/views.py]: For this limited api, the Node model is not updated. (Comments above). Technical debt issue perhaps.

**Comment by jkmarx** [comment]
@scottx611x I made some changes to this pull request including adding the datafile index and updating the UI for it. Would you please re-review it?

File: refinery/core/views.py [hunk 5]
```diff
@@ -754,20 +745,33 @@ def get(self, request, uuid):
...
+                    )
+                except (FileStoreItem.DoesNotExist,
+                        FileStoreItem.MultipleObjectsReturned) as e:
+                    logger.error(e)
+                else:
+                    file_store_item.delete_datafile()
+
+            NodeIndex().update_object(node, using="data_set_manager")
+            return Response(
+                NodeSerializer(node).data, status=status.HTTP_202_ACCEPTED
```
Comment (scottx611x) [inline refinery/core/views.py]: I think an `HTTP 200` would be more appropriate here, `202` would be the case if we called a celery task or something else async

File: refinery/data_set_manager/search_indexes.py [hunk 6]
```diff
@@ -153,6 +154,7 @@ def prepare(self, node):
 
         filetype = "" if file_store_item is None else file_store_item.filetype
         data.update({
+            NodeIndex.DATAFILE: node.get_file_store_item().datafile,
```
Comment (scottx611x) [inline refinery/data_set_manager/search_indexes.py]: I think this is troublesome for many `Nodes`. The majority of `Nodes` don't have `FileStoreItems`, which would yield `AttributeErrors` when trying to access `.datafile` on a `None`.

I think adding a few tests to `data_set_manager.tests.NodeIndexTests` regarding what `NodeIndex.DATAFILE` should contain based on at least the two scenarios (`Node` has/doesn't have a `FileStoreItem`) would be a good idea

**Comment by jkmarx** [comment]
@scottx611x Can you please take another look?

File: refinery/core/views.py [hunk 7]
```diff
@@ -754,20 +745,33 @@ def get(self, request, uuid):
...
+                    file_store_item = FileStoreItem.objects.get(
+                        uuid=node.file_uuid
+                    )
+                except (FileStoreItem.DoesNotExist,
+                        FileStoreItem.MultipleObjectsReturned) as e:
+                    logger.error(e)
+                else:
+                    file_store_item.delete_datafile()
+
+            NodeIndex().update_object(node, using="data_set_manager")
```
Comment (scottx611x) [inline refinery/core/views.py]: With the changes I just pushed to `scottx611x/pending_import_state_for_ui_datafile_addition` you could do: `node.update_solr_index()` here instead.

File: refinery/data_set_manager/search_indexes.py [hunk 8]
```diff
@@ -151,8 +152,11 @@ def prepare(self, node):
             if type(value) is set:
                 data[key] = " + ".join(str(i) for i in sorted(value))
 
+        datafile = "" if file_store_item is None else file_store_item.datafile
```
Comment (scottx611x) [inline refinery/data_set_manager/search_indexes.py]: Solr is expecting strings for index preparation,  but `FileStoreItem.datafile` returns `FieldFile` instances. 

I believe you would want: `datafile = file_store_item.datafile.name` here instead

File: refinery/core/views.py (hunk 7 above)
Comment (jkmarx) [reply refinery/core/views.py]: I'll update once we merge your pull request.

**Pull request merged**